src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,415 @@
+  has_top (homogeneous_ideal 𝒜) :=
+⟨⟨⊤, λ _ _ _, by simp only [submodule.mem_top]⟩⟩
+
+@[simp] lemma homogeneous_ideal.eq_top_iff (I : homogeneous_ideal 𝒜) : I = ⊤ ↔ I.1 = ⊤ :=
+subtype.ext_iff
+
+instance homogeneous_ideal.order : partial_order (homogeneous_ideal 𝒜) :=
+partial_order.lift _ subtype.coe_injective
+
+instance homogeneous_ideal.has_mem : has_mem A (homogeneous_ideal 𝒜) :=
+{ mem := λ r I, r ∈ I.1 }
+
+variables {𝒜}
+
+lemma ideal.is_homogeneous.inf {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊓ J).is_homogeneous 𝒜 :=
+λ i r hr, ⟨HI _ hr.1, HJ _ hr.2⟩
+
+lemma ideal.is_homogeneous.Inf {ℐ : set (ideal A)} (h : ∀ I ∈ ℐ, ideal.is_homogeneous 𝒜 I) :
+  (Inf ℐ).is_homogeneous 𝒜 :=
+begin
+  intros i x Hx, simp only [ideal.mem_Inf] at Hx ⊢,
+  intros J HJ,
+  exact h _ HJ _ (Hx HJ),
+end
+
+lemma ideal.is_homogeneous.mul {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I * J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  rw ideal.span_mul_span',
+  refine ⟨s₁ * s₂, congr_arg _ _⟩,
+  exact (set.image_mul (submonoid.subtype _).to_mul_hom).symm,
+end
+
+lemma ideal.is_homogeneous.sup {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊔ J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  refine ⟨s₁ ∪ s₂, _⟩,
+  rw [set.image_union],
+  exact (submodule.span_union _ _).symm,
+end
+
+lemma ideal.is_homogeneous.Sup
+  {ℐ : set (ideal A)} (Hℐ : ∀ (I ∈ ℐ), ideal.is_homogeneous 𝒜 I) :
+  (Sup ℐ).is_homogeneous 𝒜 :=
+begin
+  simp_rw ideal.is_homogeneous.iff_exists at Hℐ ⊢,
+  choose 𝓈 h𝓈 using Hℐ,
+  refine ⟨⋃ I hI, 𝓈 I hI, _⟩,
+  simp_rw [set.image_Union, ideal.span_Union, Sup_eq_supr],
+  conv in (ideal.span _) { rw ←h𝓈 i x },
+end
+
+variables (𝒜)
+
+instance : has_inf (homogeneous_ideal 𝒜) :=
+{ inf := λ I J, ⟨I ⊓ J, I.prop.inf J.prop⟩ }
+
+instance : has_Inf (homogeneous_ideal 𝒜) :=
+{ Inf := λ ℐ, ⟨Inf (coe '' ℐ), ideal.is_homogeneous.Inf $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_sup (homogeneous_ideal 𝒜) :=
+{ sup := λ I J, ⟨I ⊔ J, I.prop.sup J.prop⟩ }
+
+instance : has_Sup (homogeneous_ideal 𝒜) :=
+{ Sup := λ ℐ, ⟨Sup (coe '' ℐ), ideal.is_homogeneous.Sup $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_mul (homogeneous_ideal 𝒜) :=
+{ mul := λ I J, ⟨I * J, I.prop.mul J.prop⟩ }
+
+instance : has_add (homogeneous_ideal 𝒜) := ⟨(⊔)⟩
+
+end operations
+
+section homogeneous_core
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A]
+variables [algebra R A] [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+lemma ideal.homogeneous_core.gc :
+  galois_connection
+    (subtype.val : homogeneous_ideal 𝒜 → ideal A)
+    (λ I, ⟨I.homogeneous_core 𝒜, I.is_homogeneous_homogeneous_core 𝒜⟩ :
+      ideal A → homogeneous_ideal 𝒜) :=
```

Comment:
I think these functions should be extracted into definitions: `homogeneous_ideal.to_ideal` and `ideal.homogeneous_core`.
I know that the latter already exists, but maybe that version should only be used in the implementation of the bundled homogeneous core. So you could tack a `'` onto the existing version.

Reply:
There's some weirdness going on with `subtype.val` because `homogenous_ideal` is defined as a subtype but isn't reducible. Probably we should either change it to `homogeneous_ideal extends ideal`, or add `@[reducible]` so that we don't have to fight against the subtype API.

Reply:
Ah, maybe it's already reducible, in which case `subtype.val : homogeneous_ideal 𝒜 → ideal A` is just `coe`

Reply:
I'm suggesting that we build more API, instead of just forcing this into the `subtype` API.

Reply:
We changed the `subtype.val`s to `coe`s everywhere. Are people now happy?